app/models/device.rb

```diff
@@ -0,0 +1,79 @@
+class Device < ApplicationRecord
+  belongs_to :user
+
+  IOS = "iOS".freeze
+  ANDROID = "Android".freeze
+
+  validates :token, uniqueness: { scope: %i[user_id platform app_bundle] }
+
+  def create_notification(title, body, payload)
+    case platform
```

Comment:
@citizen428 I'm usually pretty wary of metaprogramming. It's often a bit too clever and makes things hard to grep for. And in this case, if this comes from `params` in the controller (disclosure: I haven't checked), someone could pass `platform: "create"` and overflow the stack. For that reason alone, I'd love to have a validation on `platform`.

Is it important to invoke some sort of failure here? If so and we raised an error or somehow returned a failure result from this method (maybe in an `else` clause on this `case`) if a value other than those two were provided, would that be enough to satisfy that?

Reply:
> I'm usually pretty wary of metaprogramming.

Fair. Dosis sola facit venenum — "only the dose makes the poison". I thought this specific case wasn't too bad.

>  if this comes from params in the controller 

Good point. I'm not sure if it does either, but the validation seems useful either way.

My main concern was really that we may silently swallow an error, which I'm generally not a big fan of. If `platform` has an unexpected value we probably want to know about it, not just ignore it.

Reply:
Great points here: I'd love to keep these platform strings as constants: it really helps for readability and being a source of truth IMHO. Definitely agree we should be validating strings.

I'll also change this code to throw an error: because this code will be executed in a asynchronous job if we don't throw an error and log it I'm afraid we'll end up in a situation where notifications aren't being delivered and we don't have any logging as to why.